File: gitdot-api/src/resource/repository.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::resource::user::UserResource;

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryResource {
    pub id: Uuid,
    pub name: String,
    pub owner: String,
    pub visibility: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub stars: u32,
    pub user_star: bool,
    pub readonly: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryPathsResource {
    pub ref_name: String,
    pub commit_sha: String,
    pub entries: Vec<RepositoryPathResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryPathResource {
    pub path: String,
    pub name: String,
    pub path_type: PathType,
    pub sha: String,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PathType {
    Blob,
    Tree,
    Commit,
    Unknown,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryBlobResource {
    pub commit_sha: String,
    pub path: String,
    pub sha: String,
    pub content: String,
    pub encoding: String,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryBlobsResource {
    pub blobs: Vec<RepositoryBlobResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryCommitsResource {
    pub commits: Vec<RepositoryCommitResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryCommitResource {
    pub owner_name: String,
    pub repo_name: String,
    pub sha: String,
    pub parent_sha: String,
    pub message: String,
    pub date: DateTime<Utc>,
    pub author: CommitAuthorResource,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub review_number: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub diff_position: Option<i32>,
    pub diffs: Vec<CommitDiffResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CommitAuthorResource {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<Uuid>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub git_name: String,
    pub email: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub image_updated_at: Option<DateTime<Utc>>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CommitDiffResource {
    pub path: String,
    pub lines_added: u32,
    pub lines_removed: u32,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryBlobPairResource {
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub old: Option<RepositoryBlobResource>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub new: Option<RepositoryBlobResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryResourcesResource {
    pub last_commit: String,
    pub last_updated: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub paths: Option<RepositoryPathsResource>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub commits: Option<RepositoryCommitsResource>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub blobs: Option<RepositoryBlobsResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum RepositoryActivityEventResource {
    Starred {
        user: UserResource,
        at: DateTime<Utc>,
    },
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryCommitFilterResource {
    pub id: Uuid,
    pub repository_id: Uuid,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub authors: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub paths: Option<Vec<String>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

File: gitdot-web/packages/gitdot-api/src/resource/repository.ts
import { z } from "zod";

import { BuildResource } from "./build";
import { QuestionResource } from "./question";
import { ReviewResource } from "./review";
import { UserResource } from "./user";

export const RepositoryFileResource = z.object({
  commit_sha: z.string(),
  path: z.string(),
  sha: z.string(),
  content: z.string(),
  encoding: z.string(),
});
export type RepositoryFileResource = z.infer<typeof RepositoryFileResource>;

export const CommitAuthorResource = z.object({
  id: z.uuid().optional(),
  name: z.string().optional(),
  git_name: z.string(),
  email: z.string(),
  image_updated_at: z.iso.datetime().nullable().optional(),
});
export type CommitAuthorResource = z.infer<typeof CommitAuthorResource>;

export const RepositoryDiffStatResource = z.object({
  path: z.string(),
  lines_added: z.number().int(),
  lines_removed: z.number().int(),
});
export type RepositoryDiffStatResource = z.infer<
  typeof RepositoryDiffStatResource
>;

export const RepositoryDiffFileResource = z.object({
  path: z.string(),
  lines_added: z.number().int(),
  lines_removed: z.number().int(),
  left_content: z.string().optional(),
  right_content: z.string().optional(),
});
export type RepositoryDiffFileResource = z.infer<
  typeof RepositoryDiffFileResource
>;

export const RepositoryCommitResource = z.object({
  owner_name: z.string(),
  repo_name: z.string(),
  sha: z.string(),
  parent_sha: z.string(),
  message: z.string(),
  date: z.iso.datetime(),
  author: CommitAuthorResource,
  review_number: z.number().int().optional(),
  diff_position: z.number().int().optional(),
  diffs: z.array(RepositoryDiffStatResource),
});
export type RepositoryCommitResource = z.infer<typeof RepositoryCommitResource>;

export const RepositoryCommitsResource = z.object({
  commits: z.array(RepositoryCommitResource),
});
export type RepositoryCommitsResource = z.infer<
  typeof RepositoryCommitsResource
>;

export const UserCommitResource = z.object({
  id: z.uuid(),
  date: z.iso.datetime(),
  redacted: z.boolean(),
  owner_name: z.string().optional(),
  repo_name: z.string().optional(),
  sha: z.string().optional(),
  parent_sha: z.string().optional(),
  message: z.string().optional(),
  author: CommitAuthorResource.optional(),
  review_number: z.number().int().optional(),
  diff_position: z.number().int().optional(),
  diffs: z.array(RepositoryDiffStatResource).default([]),
});
export type UserCommitResource = z.infer<typeof UserCommitResource>;

export const PathType = z.enum(["blob", "tree", "commit", "unknown"]);
export type PathType = z.infer<typeof PathType>;

export const RepositoryPathResource = z.object({
  path: z.string(),
  name: z.string(),
  path_type: PathType,
  sha: z.string(),
});
export type RepositoryPathResource = z.infer<typeof RepositoryPathResource>;

export const RepositoryPathsResource = z.object({
  ref_name: z.string(),
  commit_sha: z.string(),
  entries: z.array(RepositoryPathResource),
});
export type RepositoryPathsResource = z.infer<typeof RepositoryPathsResource>;

export const RepositoryFolderResource = z.object({
  type: z.literal("folder"),
  commit_sha: z.string(),
  path: z.string(),
  entries: z.array(RepositoryPathResource),
});
export type RepositoryFolderResource = z.infer<typeof RepositoryFolderResource>;

export const RepositoryBlobResource = z.discriminatedUnion("type", [
  RepositoryFileResource.extend({ type: z.literal("file") }),
  RepositoryFolderResource,
]);
export type RepositoryBlobResource = z.infer<typeof RepositoryBlobResource>;

export const RepositoryBlobsResource = z.object({
  blobs: z.array(RepositoryBlobResource),
});
export type RepositoryBlobsResource = z.infer<typeof RepositoryBlobsResource>;

export const RepositoryBlobPairResource = z.object({
  path: z.string(),
  old: RepositoryBlobResource.optional(),
  new: RepositoryBlobResource.optional(),
});
export type RepositoryBlobPairResource = z.infer<
  typeof RepositoryBlobPairResource
>;

export const RepositoryBlobDiffsResource = z.object({
  diffs: z.record(z.string(), RepositoryDiffFileResource),
});
export type RepositoryBlobDiffsResource = z.infer<
  typeof RepositoryBlobDiffsResource
>;

export const RepositoryResource = z.object({
  id: z.uuid(),
  name: z.string(),
  owner: z.string(),
  visibility: z.string(),
  description: z.string().optional(),
  stars: z.number().int().nonnegative(),
  user_star: z.boolean(),
  readonly: z.boolean(),
  created_at: z.iso.datetime(),
});
export type RepositoryResource = z.infer<typeof RepositoryResource>;

export const RepositoryQuestionsResource = z.object({
  questions: z.array(QuestionResource),
});
export type RepositoryQuestionsResource = z.infer<
  typeof RepositoryQuestionsResource
>;

export const RepositoryReviewsResource = z.object({
  reviews: z.array(ReviewResource),
});
export type RepositoryReviewsResource = z.infer<
  typeof RepositoryReviewsResource
>;

export const RepositoryBuildsResource = z.object({
  builds: z.array(BuildResource),
});
export type RepositoryBuildsResource = z.infer<typeof RepositoryBuildsResource>;

export const RepositoryResourcesResource = z.object({
  last_commit: z.string(),
  last_updated: z.iso.datetime().optional(),
  paths: RepositoryPathsResource.optional(),
  commits: RepositoryCommitsResource.optional(),
  blobs: RepositoryBlobsResource.optional(),
  questions: RepositoryQuestionsResource.optional(),
  reviews: RepositoryReviewsResource.optional(),
  builds: RepositoryBuildsResource.optional(),
});
export type RepositoryResourcesResource = z.infer<
  typeof RepositoryResourcesResource
>;

export const RepositoryActivityEventResource = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("starred"),
    user: UserResource,
    at: z.iso.datetime(),
  }),
]);
export type RepositoryActivityEventResource = z.infer<
  typeof RepositoryActivityEventResource
>;

export const RepositoryCommitFilterResource = z.object({
  id: z.uuid(),
  repository_id: z.uuid(),
  name: z.string(),
  authors: z.array(z.string()).optional(),
  tags: z.array(z.string()).optional(),
  paths: z.array(z.string()).optional(),
  created_at: z.iso.datetime(),
  updated_at: z.iso.datetime(),
});
export type RepositoryCommitFilterResource = z.infer<
  typeof RepositoryCommitFilterResource
>;
